Activity support (#161)
temporalio/sdk-ruby: README.md

**Contents**

- [Quick Start](#quick-start)
  - [Installation](#installation)
  - [Implementing an Activity](#implementing-an-activity)
  - [Running a Workflow](#running-a-workflow)
- [Usage](#usage)
  - [Client](#client)
    - [Cloud Client Using mTLS](#cloud-client-using-mtls)
    - [Data Conversion](#data-conversion)
      - [ActiveRecord and ActiveModel](#activerecord-and-activemodel)
  - [Workers](#workers)
  - [Workflows](#workflows)
  - [Activities](#activities)
    - [Activity Definition](#activity-definition)
    - [Activity Context](#activity-context)
    - [Activity Heartbeating and Cancellation](#activity-heartbeating-and-cancellation)
    - [Activity Worker Shutdown](#activity-worker-shutdown)
    - [Activity Concurrency and Executors](#activity-concurrency-and-executors)
    - [Activity Testing](#activity-testing)
- [Development](#development)
  - [Build](#build)
  - [Testing](#testing)
  - [Code Formatting and Type Checking](#code-formatting-and-type-checking)
  - [Proto Generation](#proto-generation)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Quick Start

### Installation

⚠️ PENDING GEM PUBLISH

> **NOTE: Due to [an issue](https://github.com/temporalio/sdk-ruby/issues/162), fibers (and `async` gem) are only

Author: This section is pending, there are several things to mention

> **NOTE: Due to [an issue](https://github.com/temporalio/sdk-ruby/issues/162), fibers (and `async` gem) are only

Author: :+1: We have the https://img.shields.io/pypi/pyversions/temporalio.svg?style=for-the-badge badge on the Python SDK README, but since there's no Ruby equivalent, I will probably add https://img.shields.io/badge/ruby-3.1%20%7C%203.2%20%7C%203.3-blue.svg?style=for-the-badge at the top of this README

> **NOTE: Due to [an issue](https://github.com/temporalio/sdk-ruby/issues/162), fibers (and `async` gem) are only

Author: Added badge

`my-task-queue` in Ruby, you can have the following Ruby script:

```ruby
require 'temporalio/activity'
require 'temporalio/cancellation'
require 'temporalio/client'
require 'temporalio/worker'

# Implementation of a simple activity
class SayHelloActivity < Temporalio::Activity
  def execute(name)
    "Hello, #{name}!"
  end
end

# Create a client
client = Temporalio::Client.connect('localhost:7233', 'my-namespace')

# Create a worker with the client and activities
worker = Temporalio::Worker.new(
  client:,
  task_queue: 'my-task-queue',
  # There are various forms an activity can take, see specific section for details.
  activities: [SayHelloActivity]
)

# Run the worker until SIGINT. This can be done in many ways, see specific
# section for details.
cancellation, cancel_proc = Temporalio::Cancellation.new
Signal.trap('INT') { cancel_proc }
worker.run(cancellation:)
```

Running that will run the worker until Ctrl+C pressed.

### Running a Workflow

Assuming that `SayHelloWorkflow` just calls this activity, it can be run like so:

```ruby
require 'temporalio/client'

# Create a client
client = Temporalio::Client.connect('localhost:7233', 'my-namespace')

# Run workflow

> result = client.execute_workflow(

Reviewer: Maybe add EOL comments to indicate what arguments are? In particular, the 'Temporal' arg is non-obvious.

Author: I don't think it's that common to have to annotate all uses of Ruby code for these arguments. Like Python and Go before it, the argument comes after the workflow but we similarly do not add EOL comments there to say what comes after the workflow is its argument. There will be many many examples and docs showing calling a workflow this way, I am unsure it's reasonable to expect those to all have comments (nor do I think this doc is special, there are going to be lots of quickstart/example docs).

)
puts "Result: #{result}"
```

This will output:

```
Result: Hello, Temporal!
```

## Usage

### Client

A client can be created and used to start a workflow or otherwise interact with Temporal. For example:

```ruby
require 'temporalio/client'

# Create a client
client = Temporalio::Client.connect('localhost:7233', 'my-namespace')

# Start a workflow
handle = client.start_workflow(
  'SayHelloWorkflow',
  'Temporal',
  id: 'my-workflow-id',
  task_queue: 'my-task-queue'
)

# Wait for result
result = handle.result
puts "Result: #{result}"
```

Notes about the above code:

* Temporal clients are not explicitly closed.
* To enable TLS, the `tls` option can be set to `true` or a `Temporalio::Client::Connection::TLSOptions` instance.
* Instead of `start_workflow` + `result` above, `execute_workflow` shortcut can be used if the handle is not needed.
* The `handle` above is a `Temporalio::Client::WorkflowHandle` which has several other operations that can be performed
  on a workflow. To get a handle to an existing workflow, use `workflow_handle` on the client.
* Clients are thread safe and are fiber-compatible (but fiber compatibility only supported for Ruby 3.3+ at this time).

> #### Cloud Client Using mTLS

Reviewer: Should also document API Key

Author: API key is not GA yet and we don't document it in Python or .NET SDK READMEs, nor do we document in TypeScript places like https://docs.temporal.io/develop/typescript/temporal-clients#connect-to-temporal-cloud. I would expect API key documentation for each language to be done across languages when available/ready.

Reviewer: Not GA, but it's Public Preview, which mandates that we should have basic documentation. And I expect it will be a few months before we add Ruby docs on docs.temporal.io. Till then, this README file is the main doc.

But I honestly don't mind. Your call.

Author: :+1: We should update all languages together, Ruby should not be special here.

before issuing actual cancellation to all still-running activities.

Worker shutdown will then wait on all activities to complete. If a long-running activity does not respect cancellation,
the shutdown may never complete.

#### Activity Concurrency and Executors

By default, activities run in the "thread pool executor" (i.e. `Temporalio::Worker::ActivityExecutor::ThreadPool`). This
default is shared across all workers and is a naive thread pool that continually makes threads as needed when none are
idle/available to handle incoming work. If a thread sits idle long enough, it will be killed.

The maximum number of concurrent activities a worker will run at a time is configured via its `tuner` option. The
default is `Temporalio::Worker::Tuner.create_fixed` which defaults to 100 activities at a time for that worker. When
this value is reached, the worker will stop asking for work from the server until there are slots available again.

In addition to the thread pool executor, there is also a fiber executor in the default executor set. To use fibers, call
`activity_executor :fiber` class method at the top of the activity class (the default of this value is `:default` which

> is the thread pool executor). Activities can only choose the fiber executor if they create and run the worker in a

Reviewer: > Activities can only choose the fiber executor if they create and run the worker in a fiber

Misconstructed sentence

Author: Can you clarify what is wrong here? This sentence is meant to be read exactly as stated, you can't choose a fiber executor if you didn't created/run the worker in a fiber. We will give a runtime error.

Reviewer: ```suggestion
Activities can only choose the fiber executor if **the worker has been created and run** in a fiber ...
```

Author: Fixed
